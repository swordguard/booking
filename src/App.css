.app {
  text-align: center;
  max-width: 1000px;
  padding: 50px 10px;
  margin: auto;
}
.flex {
  display: flex;
}
.flex-space-between {
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.flex-center {
  display: flex;
  justify-content: center;
}

.centered {
  text-align: center;
}
.left {
  text-align: left;
}
.bold {
  font-weight: bold;
}
.progress-bar {
  margin: 30px 0;
}
.current-step {
  background-color: gray;
}

.arrow-right {
  width: 0; 
  height: 0; 
  border-top: 15px solid transparent;
  border-bottom: 15px solid transparent;
  border-left: 15px solid green;
}

.stepbar {
  border: 1px solid gray;
  width: 33%;
  max-width: 300px;
  padding: 10px;
}

.session-container {
  width: 100%;
  margin-bottom: 20px;
}
@media screen and (min-width:768px) {
  .session-container {
    width: 600px;
  }
  .app {
    padding: 50px 50px;
  }
  .name-field {
    display: flex;
  }
  .first-name {
    width: 30%;
    margin-right: 5%;
  }
  .last-name {
    width: 65%;
  }
}


.session {
  /* max-width: 500px; */
  border: 1px solid black;
  padding: 20px;
}

fieldset {
  border: none;
}

.strikethrough, .disabled-radio {
  text-decoration: line-through
}

div.react-datepicker__day
{
  width: 3rem;
}
.form-body {
  flex-direction: column;
  align-items: center;
  padding: 30px;
    text-align: left;
}

/* Step 2 */
.text-input {
  padding: 5px 10px;
  width: 95%;
  border-radius: 3px;
  border: 1px solid gray;
}

.demographics {
  padding: 10px 0;
}

.line-input {
  margin: 20px 0;
}

.field-label {
  font-weight: 500;
}

.error {
  color: red;
}

.line-input-error {
  background-color: rgb(238, 195, 195);
  border-radius: 3px;
  padding: 5px 10px;
}
