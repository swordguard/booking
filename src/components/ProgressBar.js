import React from  'react'
import steps from '../configs/steps'

// const ArrowRight = () => {
//     return <div class="arrow-right"></div>
// }

const StepBar = ({label, isFirstStep, isLastStep, isCurrent, id, stepId}) => {
    const classes = (id === stepId ? 'current-step' : '') +  ' centered stepbar'
    return (
        <span className={classes}>
            {/* {!isFirstStep && <ArrowRight />} */}
            <div>{label}</div>
            {/* {!isLastStep && <ArrowRight />} */}
        </span>
    )
}

const renderProgressBar = ({stepId}) => {
    return steps.map(step => {
        const  {id} = step
        return <StepBar key={id} {...step} stepId={stepId}/>
    })
}
const ProgressBar = ({stepId}) => {
    return <div className='progress-bar flex-center'>
        {renderProgressBar({stepId})}
    </div>
       
}

export default ProgressBar