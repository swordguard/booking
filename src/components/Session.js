import React, {useState} from  'react'
import DatePicker from "react-datepicker";
import 'react-datepicker/dist/react-datepicker-cssmodules.css';

const MyDatePicker = () => {
    const [startDate, setStartDate] = useState(new Date());
    return (
      <DatePicker selected={startDate} onChange={(date) => setStartDate(date)} />
    );
};

const SessionTitle = ({sessionName, price}) => {
    
    return (
        <div >
                <div>{sessionName}</div>
                <div>30 minutes @ ${price}.00</div>
            
        </div>
    )
}

const DatePick = () => {
    return <MyDatePicker />
}

const Session = ({sessionName, price}) => {
    const [showDownArrow, setShowDownArrow] = useState(false);
    const onSessionTittleClick = () => {
        setShowDownArrow(!showDownArrow)
    }
    return <div onClick={onSessionTittleClick}>
        <div  className='flex-space-between'>
            <SessionTitle sessionName={sessionName} price={price}/>
            {showDownArrow && <div>\/</div>}
        </div>
        {showDownArrow &&<DatePick />}
    </div>
}

export default Session