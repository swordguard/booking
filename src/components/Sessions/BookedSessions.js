import {getTimeLabelValue, formatDate} from '../../utils'

export const BookedSessions = (props) => {
    const {state} = props
    return Object.keys(state).filter(key => state[key])
        .map(key => {
            const valueObj = state[key]
            const timeSlot = Object.keys(valueObj)[0]
            const sessionName = valueObj[timeSlot]
            const time = getTimeLabelValue(timeSlot)
            const date = `${formatDate(new Date(Number(key)))} ${time}` 
            return <div key={key+timeSlot}>{sessionName} {date}</div>
            return {
                sessionName,
                date,
                time,
            }
        })
}

export default BookedSessions