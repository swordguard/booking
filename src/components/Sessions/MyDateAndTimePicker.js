import React, {useEffect, useState} from  'react'
import DatePicker from "react-datepicker";
import {formatDateToNumber, availableTimeSlots, getTimeLabelValue, formatDate} from '../../utils'

const TimeRangeSelection = ({onTimeSelect, timeSlot, disabledSlots = []}) => {
    return (
        <div>
            <div>Please select a time</div>
            <div>
            <fieldset id="group">
                {
                    Array(availableTimeSlots).fill().map((v, i) => {
                        const change= (e) => onTimeSelect()(e.target.value, i)
                        const timeLabel = getTimeLabelValue(i)
                        const disabled = disabledSlots.includes(String(i))
                        return <div key={i} className={disabled ? 'disabled-radio' : ''}>
                                {Number(timeSlot) === i ?
                                    <input type='radio' key={i} id={i} checked  onChange={change} name="group" value={timeLabel}/> :
                                    <input type='radio' key={i} id={i} disabled={disabled} onChange={change} name="group" value={timeLabel}/>    
                                }
                                <label htmlFor={i}>{timeLabel}</label>
                        </div>
                    })
                }
                </fieldset>
            </div>
        </div>
    )
}


const MyDatePicker = ({startDate, onDatePickerChange}) => {
    return (
      <DatePicker selected={startDate} onChange={(date) => onDatePickerChange(date)} 
        inline 
        minDate={new Date()}
      />
    );
};

const MyDateAndTimePicker = ({ occupiedSlots, scheduleAnotherState, sessionName, onDateSelect, onTimeSelect, dispatch, bookedSession }) => {
    const [startDate, setStartDate] = useState(bookedSession && bookedSession.sessionName === sessionName
         ? new Date(Number(bookedSession.bookedDate)) : new Date());
    const [timeSlot, setTimeSlot] = useState();
    const [disabledSlots, setDisabledSlots] = useState([]);
    const onDatePickerChange = (dateSelected) => {
        onDateSelect(dateSelected)
        setStartDate(dateSelected)
    }

    const onTimeSelectWithDate = (date) => (label, index) => {
        // remove prepopulated timeSlot
        setTimeSlot()
        onTimeSelect({selectedDate: formatDateToNumber(date), label, index})
    }

    if (scheduleAnotherState && scheduleAnotherState.history) {

    }
    useEffect(() => {
        if (bookedSession) {
            if (bookedSession.sessionName === sessionName) {
                setTimeSlot(bookedSession.timeSlot)
            }
        }
    }, [])

    useEffect(() => {
        // [{1629022271684: "0"}, {1629022272344: "1"}]
        if (occupiedSlots) {
            const filteredSlotsDates = occupiedSlots.filter(slot => {
                const date = Object.keys(slot)[0]
                const formattedDate = new Date(Number(date))
                const occupiedYear = formattedDate.getFullYear()
                const occupiedMonth = formattedDate.getMonth()
                const occupiedDate = formattedDate.getDate()

                const formattedSelectedDate = startDate
                const selectedYear = formattedSelectedDate.getFullYear()
                const selectedMonth = formattedSelectedDate.getMonth()
                const selectedDate = formattedSelectedDate.getDate()

                return `${occupiedYear}${occupiedMonth}${occupiedDate}` === `${selectedYear}${selectedMonth}${selectedDate}`
            })
            if (filteredSlotsDates.length > 0) {
                const updatedDisabledSlots = filteredSlotsDates.map(slot => Object.values(slot)[0])
                setDisabledSlots(updatedDisabledSlots)
            } else {
                setDisabledSlots([])
            }
        }
    }, [occupiedSlots, startDate])

    // console.log('disabledSlots in render', occupiedSlots, disabledSlots)
    
    return (
        <>
            <MyDatePicker onDatePickerChange={onDatePickerChange} dispatch={dispatch} startDate={startDate}/>
            <TimeRangeSelection onTimeSelect={() => onTimeSelectWithDate(startDate)} timeSlot={timeSlot} disabledSlots={disabledSlots}/>  
        </>
    )
}

export default MyDateAndTimePicker