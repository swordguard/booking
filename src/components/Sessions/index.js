export * from './Session'
export * from './MyDateAndTimePicker'
export * from './BookedSessions'