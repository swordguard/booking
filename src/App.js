import { useState, useReducer } from 'react';
import './App.css';
import ProgressBar from './components/ProgressBar';
import steps from './configs/steps';
import {sessionReducer, demographicReducer, scheduleAnotherReducer} from './reducers';

const NavigationButton = ({onNextClick, label, disable}) => {
  return <button disable={String(disable)} onClick={onNextClick}>{label}</button>
}

const BackButton = ({onBackClick, label}) => {
  return <button onClick={onBackClick}>{label}</button>
}

function App() {
  const [stepId, setStepId] = useState(1)
  const [formValid, setFormValid] = useState(false)
  const [state, dispatch] = useReducer(sessionReducer, {})
  const [demographicState, demoDispatch] = useReducer(demographicReducer, {})
  const [scheduleAnotherState, scheduleAnotherDispatch] = useReducer(scheduleAnotherReducer, {})
  const targetStep = steps.find(step => step.id === stepId)
  const {navigatorButtonLabel, backButtonLabel, isLastStep, component} = targetStep || {}
  const CurrentStepComponent = component
  const onNextClick = () => setStepId(isLastStep ? 1 : stepId + 1)
  const onBackClick = () => setStepId(stepId - 1)
  return (
    <div className='app centered'>
      <h1>Book a wellness session.</h1>
      <div>Visit one of our expert consultants to get yourself feeling 100% again.</div>
      <ProgressBar stepId={stepId}/>
      {backButtonLabel && <BackButton onBackClick={onBackClick} label={backButtonLabel}/>}
      <CurrentStepComponent setFormValid={setFormValid} dispatch={dispatch} demoDispatch={demoDispatch} scheduleAnotherDispatch={scheduleAnotherDispatch} state={state} demographicState={demographicState} scheduleAnotherState={scheduleAnotherState} setStepId={setStepId} onBackClick={onBackClick}/>
      <NavigationButton disable={!formValid} onNextClick={onNextClick} label={navigatorButtonLabel}/>
    </div>
  );
}

export default App;
