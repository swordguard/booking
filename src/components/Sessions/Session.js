import { useState, useEffect, useContext } from "react";

import "react-datepicker/dist/react-datepicker.css";
import sessions from "../../configs/sessions";
import MyDateAndTimePicker from "./MyDateAndTimePicker";
import { ContextProvider } from "../../provider";

import { formatDateToNumber, checkIfSessionIsBooked } from "../../utils";

const SessionTitle = ({ sessionName, price }) => {
  return (
    <div className="left">
      <div className="bold">{sessionName}</div>
      <div>30 minutes @ ${price}.00</div>
    </div>
  );
};

const Session = ({
  openedSession,
  setOpenedSession,
  bookedSession,
  occupiedSlots,
  name,
  price,
  onDateSelect,
  onTimeSelect,
  // dispatch,
}) => {
  const [showDownArrow, setShowDownArrow] = useState(false);
  const onSessionTittleClick = () => {
    setShowDownArrow(!showDownArrow);
    setOpenedSession(name);
  };
  const onSessionTimeSelect =
    (name) =>
    ({ selectedDate, label, index }) => {
      onTimeSelect({ sessionName: name, selectedDate, label, index });
    };

  useEffect(() => {
    setShowDownArrow(openedSession === name);
  }, [openedSession]);

  return (
    <div className="session-container">
      <div
        onClick={onSessionTittleClick}
        className="flex-space-between session"
      >
        <SessionTitle sessionName={name} price={price} />
        {showDownArrow && <div className="flex-center">\/</div>}
      </div>
      {showDownArrow && (
        <MyDateAndTimePicker
          bookedSession={bookedSession}
          occupiedSlots={occupiedSlots}
          onDateSelect={onDateSelect}
          onTimeSelect={onSessionTimeSelect(name)}
          // dispatch={dispatch}
          // state={state}
          sessionName={name}
        />
      )}
    </div>
  );
};

export const Sessions = ({ dispatch, scheduleAnotherState, stepName }) => {
  const sessionContext = useContext(ContextProvider)[stepName];
  const [openedSession, setOpenedSession] = useState();
  const [occupiedSlots, setOccupiedSlots] = useState();
  const onTimeSelect = ({ sessionName, label, index, selectedDate }) => {
    dispatch({
      type: "UPDATE_TIME",
      stepName,
      payload: {
        sessionName,
        date: formatDateToNumber(selectedDate),
        timeSlot: index,
      },
    });
  };
  const onDateSelect = (date) => {
    dispatch({
      type: "UPDATE_DATE",
      stepName,
      payload: formatDateToNumber(date),
    });
  };

  const bookedSession = checkIfSessionIsBooked(sessionContext)[0];
  useEffect(() => {
    if (bookedSession) {
      setOpenedSession(bookedSession.sessionName);
    }
  }, []);

  useEffect(() => {
    if (scheduleAnotherState && scheduleAnotherState.history) {
      const historyState = scheduleAnotherState.history;
      const occupiedSessions = checkIfSessionIsBooked(historyState);
      if (Array.isArray(occupiedSessions)) {
        const slots = occupiedSessions.map(({ bookedDate, timeSlot }) => ({
          [bookedDate]: timeSlot,
        }));
        setOccupiedSlots(slots);
      }
    }
  }, []);

  return (
    <div className="flex-center form-body">
      {sessions.map((session) => {
        const { id } = session;
        return (
          <Session
            key={id}
            {...session}
            openedSession={openedSession}
            setOpenedSession={setOpenedSession}
            bookedSession={bookedSession}
            occupiedSlots={occupiedSlots}
            onTimeSelect={onTimeSelect}
            onDateSelect={onDateSelect}
            dispatch={dispatch}
          />
        );
      })}
    </div>
  );
};
