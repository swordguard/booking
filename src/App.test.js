import { render, screen } from '@testing-library/react';
import { unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import App from './App';
import {Sessions} from './components/Sessions'

test('renders learn react link', () => {
  render(<App />);
  const linkElement = screen.getByText(/Book a wellness session./i);
  expect(linkElement).toBeInTheDocument();
});


let container = null;
beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

it("should render a Session", () => {
  const props = {
    state: {},
    dispatch: jest.mock(),
    scheduleAnotherState: {}
  }
  act(() => {
    render(<Sessions {...props}/>, container);
  });

  expect(
    container.innerHTML
  ).toMatchSnapshot();

});
