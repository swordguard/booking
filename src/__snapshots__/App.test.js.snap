// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a Session 1`] = `""`;
